lightning/src/chain/mod.rs

```diff
@@ -9,6 +9,117 @@
 
 //! Structs and traits which allow other parts of rust-lightning to interact with the blockchain.
 
+use bitcoin::blockdata::script::Script;
+use bitcoin::blockdata::transaction::TxOut;
+use bitcoin::hash_types::{BlockHash, Txid};
+
+use chain::channelmonitor::{ChannelMonitor, ChannelMonitorUpdate, ChannelMonitorUpdateErr, MonitorEvent};
+use chain::keysinterface::ChannelKeys;
+use chain::transaction::OutPoint;
+
 pub mod chaininterface;
+pub mod chainmonitor;
+pub mod channelmonitor;
 pub mod transaction;
 pub mod keysinterface;
+
+/// The `Access` trait defines behavior for accessing chain data and state, such as blocks and
+/// UTXOs.
+pub trait Access: Send + Sync {
+	/// Returns the transaction output of a funding transaction encoded by [`short_channel_id`].
+	/// Returns an error if `genesis_hash` is for a different chain or if such a transaction output
+	/// is unknown.
+	///
+	/// [`short_channel_id`]: https://github.com/lightningnetwork/lightning-rfc/blob/master/07-routing-gossip.md#definition-of-short_channel_id
+	fn get_utxo(&self, genesis_hash: &BlockHash, short_channel_id: u64) -> Result<TxOut, AccessError>;
+}
+
+/// An error when accessing the chain via [`Access`].
+///
+/// [`Access`]: trait.Access.html
+#[derive(Clone)]
+pub enum AccessError {
+	/// The requested chain is unknown.
+	UnknownChain,
+
+	/// The requested transaction doesn't exist or hasn't confirmed.
+	UnknownTx,
+}
+
+/// The `Watch` trait defines behavior for watching on-chain activity pertaining to channels as
+/// blocks are connected and disconnected.
+///
+/// Each channel is associated with a [`ChannelMonitor`]. Implementations of this trait are
+/// responsible for maintaining a set of monitors such that they can be updated accordingly as
+/// channel state changes and HTLCs are resolved on chain. See method documentation for specific
+/// requirements.
+///
+/// Implementations **must** ensure that updates are successfully applied and persisted upon method
+/// completion. If an update fails with a [`PermanentFailure`], then it must immediately shut down
+/// without taking any further action such as persisting the current state.
+///
+/// If an implementation maintains multiple instances of a channel's monitor (e.g., by storing
+/// backup copies), then it must ensure that updates are applied across all instances. Otherwise, it
```

Comment:
This contradicts a discussion on the 681: https://github.com/rust-bitcoin/rust-lightning/pull/681#discussion_r484259489 where Antoine/devrandom concluded that the majority is safe for backups. Not sure whose should be updated though (I don't feel strongly about majority vs. all instances for v1).

Reply:
It's really a two-layer deployment : 

* at the monitor-level, you need consensus of all your monitor instance to move the off-chain state forward
* at the data-level, you need only 1 not -defaulting shard, a majority means a higher level of reliability

You may have Alice, the off-chain channel manager requiring from Bob, Dave and Caroll, its monitors, to acknowledge the state update and then Bob may ask internally to Eve and Fanny its shards to store the data before replying to Alice.

Yes this starts to be a fancy setup but this is what you want to pour confidently double-digit of BTCs in your channels :) ?